aten/src/ATen/native/mkldnn/Conv.cpp

```diff
@@ -66,377 +98,362 @@ static std::vector<int64_t> conv_output_size(
   return output_size;
 }
 
-at::Tensor mkldnn_convolution(
-    const at::Tensor& input, const at::Tensor& weight, const at::Tensor& bias,
-    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups)
-{
-  auto output = at::empty(conv_output_size(
-    input.sizes(), weight.sizes(), padding, stride, dilation, groups), input.options());
+static std::vector<int64_t> conv_input_size(
+    IntArrayRef output_size, IntArrayRef weight_size, IntArrayRef padding,
+    IntArrayRef output_padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups) {
 
-  auto cpu_engine = CpuEngine::Instance().get_engine();
-
-  int32_t g = groups;
-
-  int32_t n = input.size(0);
-  int32_t ic = input.size(1);
-  int32_t ih = input.size(2);
-  int32_t iw = input.size(3);
-
-  int32_t oc = output.size(1);
-  int32_t oh = output.size(2);
-  int32_t ow = output.size(3);
-
-  int32_t kh = weight.size(2);
-  int32_t kw = weight.size(3);
-
-  int32_t sh = stride[0];
-  int32_t sw = stride[1];
-  int32_t ph = padding[0];
-  int32_t pw = padding[1];
-
-  auto data_t = memory::data_type::f32;
-  auto format_any = memory::format::any;
-  auto format_nchw = memory::format::nchw;
-  auto format_weight = (g!= 1) ? memory::format::goihw : memory::format::oihw;
-  auto format_x = memory::format::x;
-
-  memory::dims input_tz = {n, ic, ih, iw};
-  memory::dims weight_tz = (g!= 1) ? memory::dims{g, oc/g, ic/g, kh, kw} : memory::dims{oc, ic, kh, kw};
-  memory::dims bias_tz = {oc};
-  memory::dims output_tz = {n, oc, oh, ow};
-  memory::dims _stride = {sh, sw};
-  memory::dims _padding = {ph, pw};
-
-  auto input_md = memory::desc({input_tz}, data_t, format_any);
-  auto weight_md = memory::desc({weight_tz}, data_t, format_any);
-  auto bias_md = memory::desc({bias_tz}, data_t, format_any);
-  auto output_md = memory::desc({output_tz}, data_t, format_any);
-
-  std::shared_ptr<convolution_forward::desc> conv_forward_desc;
-  if (bias.defined()) {
-    conv_forward_desc.reset(new convolution_forward::desc(prop_kind::forward,
-      convolution_direct, input_md, weight_md, bias_md, output_md,
-      _stride, _padding, _padding, padding_kind::zero));
-  } else {
-    conv_forward_desc.reset(new convolution_forward::desc(prop_kind::forward,
-      convolution_direct, input_md, weight_md, output_md,
-      _stride, _padding, _padding, padding_kind::zero));
+  auto dim = output_size.size();
+  std::vector<int64_t> input_size(dim);
+  input_size[0] = output_size[output_batch_size_dim];
+  input_size[1] = weight_size[weight_output_channels_dim] * groups;
+
+  for (size_t d = 2; d < dim; ++d) {
+    auto kernel = dilation[d - 2] * (weight_size[d] - 1) + 1;
+    input_size[d] = (output_size[d] - 1) * stride[d - 2] - (2 * padding[d - 2])
+                        + kernel + output_padding[d - 2];
+  }
+  return input_size;
+}
+
+struct ConvolutionParams {
```

Comment:
We shouldn't cargo cult this struct here. ConvolutionParams in cudnn is a POD struct because we need to hash it to maintain our benchmark cache. There is no such for mkldnn, right?

Reply:
Indeed, it looks like most stuff has the right memory layout already, so there's no need to stuff it in ConvolutionParams, just pass it along directly.

Reply:
The reason to separate `ConvolutionParams` and `ConvolutionArgs` here is exactly same as cuDNN. `mkldnn` also has several convolution algorithms, `direct` or `winograd`. At this moment only `direct` is used, but i am planning to add `winograd`  later on and do the algorithm cache. So `ConvolutionParams` is a POD struct used for as a key for caching the algorithm and `ConvolutionArgs` is the one holding `mkldnn` attributes which is used to register mkldnn's memory descriptor.